cmake/template_instantiation.cmake

```diff
@@ -0,0 +1,68 @@
+function(add_instantiation_files source_dir source_file output_files_var)
```

Comment:
maybe instead of an output variable, we could pass in the target the sources should be added to. Then we would not need to add them manually.

Reply:
That wouldn't be sufficient for HIP and CUDA, where we need to set some properties on the files on top of adding them to a target.

Reply:
What do you mean by that? At least I can't find another place where the output variable is used except for `target_sources`.

Reply:
https://github.com/ginkgo-project/ginkgo/blob/dbd43cf0624fba5fb8fe7d8850955180f0e1e7c6/cuda/CMakeLists.txt#L60

Reply:
I guess we could get the files from the target's `SOURCES` property. But maybe that is not worth the trouble.

Reply:
Yes, with native HIP support we could consider it